### PR TITLE
Add DD4Hep version MF producer from DB

### PR DESCRIPTION
**Pull request by namapane**
#### PR description:
Add DD4Hep version of MF producer from DB. This completes the  migration started with #27886. 
(the default MF producer is unchanged; switch to the DD4Hep version will happen at a later stage.)

Also, add options to test this producer in the regression test.

#### PR validation:
Ran regression test against reference files with all combination of MF producer, era and current.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+#include "CondFormats/DataRecord/interface/MagFieldConfigRcd.h"
+
+#include <string>
+#include <vector>
+#include <iostream>
+#include <memory>
+
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/lexical_cast.hpp>
```
Comment (slava77) [inline MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc]: I do not see any boost:: in this file; can these includes be dropped?

File: MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  // Build the geometry from the DB blob
+
+  auto const& blob = iRecord.getTransientHandle(mayConsumeBlobToken_);
+  std::unique_ptr<std::vector<unsigned char> > tb = blob->getUncompressedBlob();
+
+  string sblob(tb->begin(), tb->end());
+  sblob.insert(
+      sblob.rfind("</DDDefinition>"),
+      "<MaterialSection label=\"materials.xml\"><ElementaryMaterial name=\"materials:Vacuum\" density=\"1e-13*mg/cm3\" "
+      "symbol=\" \" atomicWeight=\"1*g/mole\" atomicNumber=\"1\"/></MaterialSection>");
```
Comment (slava77) [inline MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc]: this looks like a kludge.
Is this insert [still] needed?

File: MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+#include "FWCore/Framework/interface/ESTransientHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/ESProductTag.h"
+
+#include "DetectorDescription/Core/interface/DDCompactView.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "MagneticField/GeomBuilder/src/DD4hep_MagGeoBuilder.h"
+
+#include "DetectorDescription/DDCMS/interface/DDCompactView.h"
```
[the remark this answers is not in the document]
Comment (namapane) [reply MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc]: Indeed, thanks for spotting.

File: MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc (hunk 1 above)
Comment (namapane) [reply MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc]: Indeed (then also from plugins/VolumeBasedMagneticFieldESProducerFromDB.cc which, is the original DD4hep version)

File: MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc (hunk 2 above)
Comment (namapane) [reply MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc]: Yes. DD4Hep expects vaccuum to be defined, which was not the case in DDD. Current DB payloads do not therefore include its definition. Adding it here is simpler and safer than replacing all DB payloads.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/28832	e6c5aaf 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences (the new DD4HEP MF is not used yet in the default workflows)

**Comment by silviodonato** [comment]
+1

**Pull request merged**